src/hotspot/share/services/heapDumper.cpp

```diff
@@ -1446,18 +1790,152 @@ void HeapObjectDumper::do_object(oop o) {
   }
 }
 
+bool HeapObjectDumper::is_large(oop o) {
+  size_t size = 0;
+  if (o->is_instance()) {
+    InstanceKlass* ik = InstanceKlass::cast(o->klass());
+    size = DumperSupport::instance_size(ik);
+  } else if (o->is_objArray()) {
+    objArrayOop array = objArrayOop(o);
+    BasicType type = ArrayKlass::cast(array->klass())->element_type();
+    assert(type >= T_BOOLEAN && type <= T_OBJECT, "invalid array element type");
+    int length = array->length();
+    int type_size = sizeof(address);
+    size = (size_t)length * type_size;
+  } else if (o->is_typeArray()) {
+    typeArrayOop array = typeArrayOop(o);
+    BasicType type = ArrayKlass::cast(array->klass())->element_type();
+    assert(type >= T_BOOLEAN && type <= T_OBJECT, "invalid array element type");
+    int length = array->length();
+    int type_size = type2aelembytes(type);
+    size = (size_t)length * type_size;
+  }
+  return size > HeapDumpLargeObjectList::LargeObjectSizeThreshold;
+}
+
+// The dumper controller for parallel heap dump
+class DumperController : public CHeapObj<mtInternal> {
+ private:
+   bool     _started;
+   Monitor* _lock;
+   uint   _dumper_number;
+   uint   _waiting_number;
```

Comment:
The _waiting_number doesn't seem to really have a purpose. No decision depends on it.